aten/src/ATen/native/mkldnn/Attention.cpp

```diff
@@ -0,0 +1,546 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Config.h>
+#include <c10/core/SymIntArrayRef.h>
+#include <c10/util/ArrayRef.h>
+#include <torch/library.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_to_dense_native.h>
+#include <ATen/ops/empty.h>
+#include <ATen/ops/empty_like.h>
+#endif
+
+#if AT_ONEDNN_GRAPH_ENABLED()
+
+#include <ATen/native/mkldnn/Graph.h>
+#include <ATen/native/mkldnn/Utils.h>
+#include <c10/util/irange.h>
+#include <omp.h>
+#include <oneapi/dnnl/dnnl_graph.hpp>
+
+namespace at {
+namespace native {
+
+namespace {
+
+using namespace onednn_graph;
+
+/*
+
+GPT2 MHA kernel
+
+Attention-mask is non-optional but would be
+optional in oneDNN v3.5 for better performance.
+
+ [query]    [key]
+      \     /
+[cond] MatMul [mask]
+     \   |   /
+       Select   [scale]
+            \   /
+             Div   [add in]
+               \   /
+                Add
+                 |
+               Softmax   [value]
+                    \     /
+                     MatMul
+                       |
+                StaticTranspose
+                       |
+                    Reorder
+                       |
+                    [output]
```

Comment:
Can we make C++ implementation not pattern-specific? I guess the new op `mkldnn::_graph_sdpa_pattern` is supposed to be general, not specific to particular patterns like GPT2?

Reply:
Thanks, as discussed offline, I'll make this part generic asap, and will add comments pertaining to each pattern.